### ISSUE TITLE
SPARTA_FLATTEN on STL types

### ISSUE THREAD

**Issue by klingaard** [sparcians/map#74]
Curious, can I use `SPARTA_FLATTEN` on STL types that contain objects with `PairDefinitions`?

**Comment by klingaard** [comment]
I currently (understandably) get this:
```
 error: no member named 'type' in 'std::__1::vector<TYPE>'
            T::type::nestedPairCallback(this, std::forward<Args>(args)...);
```

**Comment by klingaard** [comment]
Be totally (totally) awesome if `SpartaPairDefinitions` had a `processIfSTLType`.

**Comment by avinabadasgupta** [comment]
This is achievable. Will look into it.

**Comment by klingaard** [comment]
Wanted to give you a prime example of this one...
```
class Foo
{
public:
     using SpartaPairDefinitionType = FooPairs;

     int getInt() const;
};

// FooPairs type defined here...

class Bar
{
public:
    using SpartaPairDefinitionType = BarPairs;
    std::vector<Foo> getFoos() const;
};

class BarPairs : public sparta::PairDefinition<Bar>
{
public:
    BarPairs() : PairDefinition<Bar>()
    {
         SPARTA_INVOKE_PAIRS(Bar);
    }
    SPARTA_REGISTER_PAIRS(SPARTA_ADDPAIR("foos", &PipeOp::getFoos));
};
```

**Comment by avinabadasgupta** [comment]
> Wanted to give you a prime example of this one...
> 
> ```
> class Foo
> {
> public:
>      using SpartaPairDefinitionType = FooPairs;
> 
>      int getInt() const;
> };
> 
> // FooPairs type defined here...
> 
> class Bar
> {
> public:
>     using SpartaPairDefinitionType = BarPairs;
>     std::vector<Foo> getFoos() const;
> };
> 
> class BarPairs : public sparta::PairDefinition<Bar>
> {
> public:
>     BarPairs() : PairDefinition<Bar>()
>     {
>          SPARTA_INVOKE_PAIRS(Bar);
>     }
>     SPARTA_REGISTER_PAIRS(SPARTA_ADDPAIR("foos", &PipeOp::getFoos));
> };
> ```

Just for the sake of completion, it would look something like this right
```
class Foo
{
public:
     using SpartaPairDefinitionType = FooPairs;

     int getInt() const;
};

class FooPairs : public sparta::PairDefinition<Foo>
{
public:
    FooPairs() : PairDefinition<Foo>()
    {
         SPARTA_INVOKE_PAIRS(Foo);
    }
    SPARTA_REGISTER_PAIRS(SPARTA_ADDPAIR("some_int", &Foo::getInt));
};

class Bar
{
public:
    using SpartaPairDefinitionType = BarPairs;
    std::vector<Foo> getFoos() const;
};

class BarPairs : public sparta::PairDefinition<Bar>
{
public:
    BarPairs() : PairDefinition<Bar>()
    {
         SPARTA_INVOKE_PAIRS(Bar);
    }
    SPARTA_REGISTER_PAIRS(SPARTA_ADDPAIR("foos", &Bar::getFoos));
};
```

**Comment by klingaard** [comment]
Yeppers!

**Comment by avinabadasgupta** [comment]
> Yeppers!

This would work even today if you provide an ostream overload for vector<Foo> class. It will fall back and use annotation method.

**Comment by klingaard** [comment]
Oh, it does work!  Just fine, actually.  The problem is in our code `Foo` is a complex object that really shouldn't be an annotation -- it doubles the size of the database.

**Comment by avinabadasgupta** [comment]
> Oh, it does work! Just fine, actually. The problem is in our code `Foo` is a complex object that really shouldn't be an annotation -- it doubles the size of the database.

Definitely, the long term solution would be to not use annotation. I guess we need to make it generic and consider all STL containers and container adaptors when designing this.

**Comment by klingaard** [comment]
Hey Avi, this might be percolating to the top of the stack.  If you have some time in the next week to take a stab at this, we'd be grateful.  It's not a blocker.

**Comment by avinabadasgupta** [comment]
> Hey Avi, this might be percolating to the top of the stack. If you have some time in the next week to take a stab at this, we'd be grateful. It's not a blocker.

Sure, I'll start working on it from tomorrow.
